atlascope/core/management/commands/populate.py

```diff
@@ -82,11 +93,15 @@ def command(password):
     for model, filename in MODEL_JSON_MAPPING:
         print('-----')
         objects = json.load(open(DATALOADER_DIR + filename))
-        for obj in objects:
-            obj, many_to_many_values, files_to_save, permissions = expand_references(obj, model)
-            db_obj = model(**obj)
+        for data_obj in objects:
+            # will raise Exception if invalid field or value in data
+            MODEL_SERIALIZERS[model](data=data_obj)
```

Comment:
Is this what you mean @mcovalt? This would prevent the creation of those objects if they cannot be instantiated as the proper model serializer.

Reply:
Yes, didn't see this. Thanks for bringing it to my attention.

Reply:
Oh, you didn't miss it, don't worry! I just added it after your suggestion!

Reply:
I see this was recently committed. Sorry for the confusion. Isn't the call to `.is_valid()` required? [See here](https://www.django-rest-framework.org/api-guide/serializers/#raising-an-exception-on-invalid-data).

Reply:
I attempted to put an incorrect field in one of the data objects and it correctly raised an exception preventing the save of that object and stopping the script. If we have `raise_exception=True` on the call to `is_valid`, wouldn't that give us the same behavior?

Reply:
Not always. Validation in Django is spread across both database constraints and Django's application logic. [Not always in ways you'd think](https://docs.djangoproject.com/en/4.0/ref/models/fields/#django.db.models.ForeignKey.on_delete) [0]. [Django's default validation framework requires the explicit call to `.clean()`](https://docs.djangoproject.com/en/4.0/ref/models/instances/#validating-objects). This is used to handle custom validation outside the scope of what is possible with database constraints (and within the scope, too [1]).

The introduction of Django REST Framework further complicates this. [Django REST Framework never actually calls the `Model.clean()` method](https://www.django-rest-framework.org/community/3.0-announcement/#differences-between-modelserializer-validation-and-modelform). [Instead, it wants all validation to happen within it](https://www.django-rest-framework.org/api-guide/validators/#validation-in-rest-framework)... Kinda [2]...

Of course, there could be validation and business logic expressed in the view, too. Or middleware. But then we're really in trouble and have business logic all over the place. Instead, it should _all_ go in the serializer so we can more easily understand the constraints placed on our data.

**tl;dr**: we'll put more validation in the serializers later and a call to `.is_valid()` will run that validation.

[0] e.g. the `ON DELETE` constraint is never actually enforced in PostgreSQL. Instead, Django emulates this.
[1] e.g. unique constraints are tested by Django _and_ enforced by the database.
[2] Some constraints have to be put on the model for correct database representation (e.g. `unique=True` on a field). Django REST Framework's `ModelSerializer` inspects the model class on instantiation to pull in these constraints. See it's [metaclass](https://github.com/encode/django-rest-framework/blob/c5be86a6dbf3d21b00a296af5994fa075826bf0b/rest_framework/serializers.py#L278-L340) and [its field building methods](https://github.com/encode/django-rest-framework/blob/c5be86a6dbf3d21b00a296af5994fa075826bf0b/rest_framework/serializers.py#L1199-L1630). It's terribly complex. The worst of it is that `ModelSerializer` is limited in capability and often you'll often need to convert a `ModelSerializer` class to a `Serializer` class. At that stage you need to explicitly add all this validation to each field. Getting that serializer's validation to match with Django's model validation is a complicated and error-prone process that is necessary due to these competing validation philosophies of Django, DRF, and Postgres.

Reply:
It would be nice if this validation occurred during a test. That way, code that changes the constraints of our data will be blocked from merging if the fixtures are made invalid with those changes.

Reply:
That's a good point, thanks for explaining. I'll add `is_valid` and move the check to a tox test.

Reply:
Changed in 2e2738a

Reply:
Great, thanks!